### PR TITLE
Fix {:#?} representation of proc_macro::Literal

### PR DESCRIPTION
**Pull request by dtolnay**
Before:

```rust
TokenStream [
    Ident {
        ident: "name",
        span: #0 bytes(37..41),
    },
    Punct {
        ch: '=',
        spacing: Alone,
        span: #0 bytes(42..43),
    },
    Literal { lit: Lit { kind: Str, symbol: "SNPP", suffix: None }, span: Span { lo: BytePos(44), hi: BytePos(50), ctxt: #0 } },
    Punct {
        ch: ',',
        spacing: Alone,
        span: #0 bytes(50..51),
    },
    Ident {
        ident: "owner",
        span: #0 bytes(56..61),
    },
    Punct {
        ch: '=',
        spacing: Alone,
        span: #0 bytes(62..63),
    },
    Literal { lit: Lit { kind: Str, symbol: "Canary M Burns", suffix: None }, span: Span { lo: BytePos(64), hi: BytePos(80), ctxt: #0 } },
]
```

After:

```rust
TokenStream [
    Ident {
        ident: "name",
        span: #0 bytes(37..41),
    },
    Punct {
        ch: '=',
        spacing: Alone,
        span: #0 bytes(42..43),
    },
    Literal {
        kind: Str,
        symbol: "SNPP",
        suffix: None,
        span: #0 bytes(44..50),
    },
    Punct {
        ch: ',',
        spacing: Alone,
        span: #0 bytes(50..51),
    },
    Ident {
        ident: "owner",
        span: #0 bytes(56..61),
    },
    Punct {
        ch: '=',
        spacing: Alone,
        span: #0 bytes(62..63),
    },
    Literal {
        kind: Str,
        symbol: "Canary M Burns",
        suffix: None,
        span: #0 bytes(64..80),
    },
]
```

### CODE REVIEW

File: src/libproc_macro/bridge/client.rs [hunk 1]
```diff
@@ -202,10 +202,15 @@ impl Clone for Literal {
     }
 }
 
-// FIXME(eddyb) `Literal` should not expose internal `Debug` impls.
 impl fmt::Debug for Literal {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str(&self.debug())
+        f.debug_struct("Literal")
+            .field("kind", &format_args!("{}", &self.debug_kind()))
```
Comment (petrochenkov) [inline src/libproc_macro/bridge/client.rs:208]:
```suggestion
            .field("kind", &self.debug_kind())
```
?

Or is it used to cancel the effect of `{:#?}`, similarly to the suffix? Could you add a comment then?

**Comment by petrochenkov** [comment]
r=me with https://github.com/rust-lang/rust/pull/72233#discussion_r425914322 clarified.

File: src/libproc_macro/bridge/client.rs (hunk 1 above)
Comment (dtolnay) [reply src/libproc_macro/bridge/client.rs:208]: `self.debug_kind()` is a String so your suggestion prints it with quotes around, which is not what I want. The {} representation prints without quotes, as in `kind: Float`.

File: src/libproc_macro/bridge/client.rs [hunk 2]
```diff
@@ -202,10 +202,16 @@ impl Clone for Literal {
     }
 }
 
-// FIXME(eddyb) `Literal` should not expose internal `Debug` impls.
 impl fmt::Debug for Literal {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_str(&self.debug())
+        f.debug_struct("Literal")
+            // format the kind without quotes, as in `kind: Float`
+            .field("kind", &format_args!("{}", &self.debug_kind()))
```
Comment (dtolnay) [inline src/libproc_macro/bridge/client.rs:209]: Added comment.

**Pull request merged**